src/disco/metrics/fd_prometheus.c

```diff
@@ -0,0 +1,147 @@
+#include "fd_prometheus.h"
+
+#include "../topo/fd_topo.h"
+
+#include "../../ballet/http/fd_http_server.h"
+
+#define PRINT_LINK_IN  (0)
+#define PRINT_LINK_OUT (1)
+#define PRINT_TILE     (2)
+
+static ulong
+find_producer_out_idx( fd_topo_t const *      topo,
+                       fd_topo_tile_t const * producer,
+                       fd_topo_tile_t const * consumer,
+                       ulong                  consumer_in_idx ) {
+  /* This finds all reliable consumers of the producers primary output,
+     and then returns the position of the consumer (specified by tile
+     and index of the in of that tile) in that list. The list ordering
+     is not important, except that it matches the ordering of fseqs
+     provided to the mux tile, so that metrics written for each link
+     index are retrieved at the same index here.
+
+     This is why we only count reliable links, because the mux tile only
+     looks at and writes producer side diagnostics (is the link slow)
+     for reliable links. */
+
+  ulong count = 0UL;
+  for( ulong i=0UL; i<topo->tile_cnt; i++ ) {
+    fd_topo_tile_t const * consumer_tile = &topo->tiles[ i ];
+    for( ulong j=0UL; j<consumer_tile->in_cnt; j++ ) {
+      if( FD_UNLIKELY( consumer_tile->in_link_id[ j ] == producer->out_link_id_primary && consumer_tile->in_link_reliable[ j ] ) ) {
+        if( FD_UNLIKELY( consumer==consumer_tile && consumer_in_idx==j ) ) return count;
+        count++;
+      }
+    }
+  }
+  return ULONG_MAX;
+}
+
+static void
+prometheus_print1( fd_topo_t const *         topo,
+                   fd_http_server_t *        http,
+                   char const *              tile_name,
+                   ulong                     metrics_cnt,
+                   fd_metrics_meta_t const * metrics,
+                   int                       print_mode ) {
+  for( ulong i=0UL; i<metrics_cnt; i++ ) {
+    fd_metrics_meta_t const * metric = &metrics[ i ];
+    fd_http_server_printf( http, "# HELP %s %s\n# TYPE %s %s\n", metric->name, metric->desc, metric->name, fd_metrics_meta_type_str( metric ) );
+
+    for( ulong j=0UL; j<topo->tile_cnt; j++ ) {
+      fd_topo_tile_t const * tile = &topo->tiles[ j ];
+      if( FD_LIKELY( tile_name!=NULL && strcmp( tile->name, tile_name ) ) ) continue;
```

Comment:
No need to fix it now, but could you add a `FIXME` that this makes printing each with `PRINT_TILE` O(n^2) instead of O(n)?